### ISSUE TITLE
FUSION_E_REF_DEF_MISMATCH when CORE_ROOT environment var is set

### ISSUE THREAD

**Issue by dagood** [dotnet/cli#1663]
## Steps to reproduce

On Windows, build the product binaries for CoreCLR. Then:

* `coreclr\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root` includes `System.Runtime.dll` (`4.6.23127.0`)
* `coreclr\Tools\dotnetcli\bin` is an extracted CLI zip (1.0.0.001504), containing `System.Runtime.dll`.(`4.6.23811.0`)

Both of these dlls are strong-name signed. `sn -T` shows they have the same public key token `b03f5f7f11d50a3a`. 

In `coreclr\` run:

```
set CORE_ROOT=bin\tests\Windows_NT.x64.Debug\Tests\Core_Root
Tools\dotnetcli\bin\dotnet.exe restore --help
```

## Actual behavior

A crash popup, allowing me to debug, and:

> Unhandled Exception: System.IO.FileLoadException: Could not load file or assembly 'System.Runtime, Version=4.0.21.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040 (FUSION_E_REF_DEF_MISMATCH))

By setting `ComPlus_ForceLog=1` before running I can get a fusion log, but I don't know what's valuable from these:

```
*** Assembly Binder Log Entry  (3/2/2016 @ 3:34:34 PM) ***

The operation failed.
Bind result: hr = 0000000080131040.

--- A detailed log follows.

=== Pre-bind state information ===
LOG: By-name Bind. Display name = System.Runtime, Version=4.0.21.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Calling assembly : unknown
===
```

## Expected behavior

The help text shows up.

## Context

I worked around this for https://github.com/dotnet/coreclr/pull/3405 (discussion https://github.com/dagood/coreclr/commit/1c9ef2f9f68ffd919f6f42364453d1341d0e3795#diff-0b192804a6349e8c26d2b027afbd89a2R98 and https://github.com/dagood/coreclr/commit/0d71a87375bfa54af10b68b74a4c8d93430ed308#diff-6178796b5428b89bb33ad7bf80143a79R77) by doing `(set CORE_ROOT=)` immediately before calling `dotnet.exe`. Setting `CORE_ROOT` to any other directory without a bunch of dlls in it also works around the issue.

It seems to me that it's wrong to have `CORE_ROOT` set when trying to restore but I want to make sure the issue is understood.

/cc @ellismg

**Comment by ellismg** [comment]
> It seems to me that it's wrong to have CORE_ROOT set when trying to restore but I want to make sure the issue is understood.

FWIW, my opinion is that it should be fine to have `CORE_ROOT` set when running `dotnet.exe`.  I don't think that CLI needs anything like this, and if it does it would be ideal if it used a different variable than whatever `corerun` is using.

**Comment by piotrpMSFT** [comment]
I just checked CLI sources and I see no mention of CORE_ROOT. I'm not familiar with this particular environment variable but it seems the behavioral difference is coming from outside CLI. 

@schellap, any ideas here?

**Comment by davidfowl** [comment]
What is core_root and what are you trying to do in the first place?

**Comment by dagood** [comment]
@piotrpMSFT I've only poked around, but the only place I've seen `CORE_ROOT` is as an environment variable that `corerun` checks. A search on CoreCLR has some interesting results: https://github.com/dotnet/coreclr/search?l=cpp&q=CORE_ROOT&type=Code&utf8=%E2%9C%93.

With that info I assumed that dotnet CLI is using corerun, but (see [second discussion link](https://github.com/dagood/coreclr/commit/0d71a87375bfa54af10b68b74a4c8d93430ed308#diff-6178796b5428b89bb33ad7bf80143a79R77)) there may be more going on here than I've seen.

@davidfowl The PR was my first time looking at CoreCLR's testing code, but as far as I can tell `CORE_ROOT` is just a place where a bunch of test dependency binaries are gathered. It's set somewhat globally in the build process, so I had to do a workaround to unset it for this specific `dotnet restore` that fails under that environment.

**Comment by dagood** [comment]
@schellap looked at this with me and it turns out that the version I was using (1.0.0.001504) used corerun in the restore command. This behavior was changed here: https://github.com/dotnet/cli/commit/834edfbc9ccbe2dbe63b89a8d09c4199d000da38#diff-09b726ef412fcd8fa1a0026451b7f324

Looking at the source using the git SHA in `.version` confirms: https://github.com/dotnet/cli/blob/388bb7260c729e27991eb7b4b91a2eab6776f168/src/dotnet/commands/dotnet-restore/NuGet3.cs

1.0.0.001504 was the LKG I used to upgrade CoreFX, WCF, and Buildtools. (That's why it's so (relatively) old.) I've confirmed that the `CORE_ROOT` error doesn't repro in the latest version.

I'll make an issue in coreclr to remove the workaround when the CLI version is upgraded.
